xskillscore/core/np_deterministic.py

```diff
@@ -116,17 +130,99 @@ def _pearson_r_p_value(a, b, weights, axis, skipna):
     """
     r = _pearson_r(a, b, weights, axis, skipna)
     a = np.rollaxis(a, axis)
-    df = a.shape[0] - 2
-    t_squared = r ** 2 * (df / ((1.0 - r) * (1.0 + r)))
-    _x = df / (df + t_squared)
+    b = np.rollaxis(b, axis)
+    # dirty fix: want to check the number of nans in union(a,b) on axis 0
+    nnans = np.sum(np.isnan(a[0] * b[0]))
+    dof = a.shape[0] - 2 - nnans
+    t_squared = r ** 2 * (dof / ((1.0 - r) * (1.0 + r)))
+    _x = dof / (dof + t_squared)
     _x = np.asarray(_x)
     _x = np.where(_x < 1.0, _x, 1.0)
-    _a = 0.5 * df
+    _a = 0.5 * dof
     _b = 0.5
     res = special.betainc(_a, _b, _x)
     return res
 
 
+def _spearman_r(a, b, weights, axis, skipna):
+    """
+    ndarray implementation of scipy.stats.spearmanr.
+
+    Parameters
+    ----------
+    a : ndarray
+        Input array.
+    b : ndarray
+        Input array.
+    axis : int
+        The axis to apply the correlation along.
+    weights : ndarray
+        Input array.
+    skipna : bool
+        If True, skip NaNs when computing function.
+
+    Returns
+    -------
+    res : ndarray
+        Spearmanr's correlation coefficient.
+
+    See Also
+    --------
+    scipy.stats.spearmanr
+
+    """
+    # TODO: returns 1 if all nan
+    if skipna:
+        rankfunc = bn.rankdata
+    else:
+        rankfunc = bn.nanrankdata
+    _a = rankfunc(a, axis=axis)
+    _b = rankfunc(b, axis=axis)
+    return _pearson_r(_a, _b, weights, axis, skipna)
+
+
+def _spearman_r_p_value(a, b, weights, axis, skipna):
+    """
+    ndarray implementation of scipy.stats.spearmanr.
+
+    Parameters
+    ----------
+    a : ndarray
+        Input array.
+    b : ndarray
+        Input array.
+    axis : int
+        The axis to apply the correlation along.
+    weights : ndarray
+        Input array.
+    skipna : bool
+        If True, skip NaNs when computing function.
+
+    Returns
+    -------
+    res : ndarray
+        2-tailed p-value.
+
+    See Also
+    --------
+    scipy.stats.spearmanr
+
+    Reference
+    ---------
+    https://github.com/scipy/scipy/blob/v1.3.1/scipy/stats/stats.py#L3613-L3764
+
+    """
+    rs = _spearman_r(a, b, weights, axis, skipna)
+    a = np.rollaxis(a, axis)
+    b = np.rollaxis(b, axis)
+    # dirty fix: want to check the number of nans in union(a,b) on axis 0
+    nnans = np.sum(np.isnan(a[0] * b[0]))
+    dof = a.shape[0] - 2 - nnans
```

Comment:
wonder if we can simply put an if statement `if np.allnan(a) ...` (or what ever it is) then return ...

Reply:
here I just want to reduce the dofs. another point raised elsewhere is whether if all datapoints are nan, the results for any correlation should be nan and for _p_value should be 1 or nan @bradyrx

Reply:
Perhaps create an additional function called `_check_for_all_nans` which returns nan if a or b is full of nans. We can then use that in all the metrics. I'm not very familiar with warnings but we could raise a warning saying "array [a/b] is full of nans returning nan".